internal/circle/client_config_test.go

```diff
@@ -0,0 +1,424 @@
+			EncryptedApiKey:    &encryptedApiKey,
+			EncrypterPublicKey: &encrypterPublicKey,
+		}
+
+		err := ccm.insert(ctx, tx, config)
+		assert.Error(t, err)
+		assert.Contains(t,
+			err.Error(),
+			"invalid circle config for insert: wallet_id, encrypted_api_key, and encrypter_public_key must be provided")
+	})
+
+	t.Run("insert fails with missing encrypter_public_key", func(t *testing.T) {
+		tx := beginTx(t, ctx, dbConnectionPool)
+		defer rollback(t, tx)
+
+		config := ClientConfigUpdate{
+			WalletID:           &walletID,
+			EncryptedApiKey:    &encryptedApiKey,
+			EncrypterPublicKey: nil,
+		}
+
+		err := ccm.insert(ctx, tx, config)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "invalid circle config for insert: wallet_id, encrypted_api_key, and encrypter_public_key must be provided")
+	})
+
+	t.Run("insert fails when inserting a second record", func(t *testing.T) {
+		tx := beginTx(t, ctx, dbConnectionPool)
+		defer rollback(t, tx)
+
+		config := ClientConfigUpdate{
+			WalletID:           &walletID,
+			EncryptedApiKey:    &encryptedApiKey,
+			EncrypterPublicKey: &encrypterPublicKey,
+		}
+
+		err := ccm.insert(ctx, tx, config)
+		require.NoError(t, err)
+
+		err = ccm.insert(ctx, tx, config)
+		assert.EqualError(t, err, "inserting circle config: pq: circle_client_config must contain exactly one row")
+	})
+}
+
+func Test_ClientConfigUpdate_Validate(t *testing.T) {
+	walletID := "wallet_id"
+	encryptedApiKey := "encrypted_api_key"
+	encrypterPublicKey := "encrypter_public_key"
+
+	tests := []struct {
+		name    string
+		input   ClientConfigUpdate
+		wantErr error
+	}{
+		{
+			name:    "both wallet_id and encrypted_api_key are nil",
+			input:   ClientConfigUpdate{},
+			wantErr: errors.New("wallet_id or encrypted_api_key must be provided"),
+		},
+		{
+			name:    "encrypted_api_key is provided without encrypter_public_key",
+			input:   ClientConfigUpdate{EncryptedApiKey: &encryptedApiKey},
+			wantErr: errors.New("encrypter_public_key must be provided if encrypted_api_key is provided"),
+		},
+		{
+			name:  "wallet_id is provided without encrypted_api_key",
+			input: ClientConfigUpdate{WalletID: &walletID},
+		},
+		{
+			name:  "both wallet_id and encrypted_api_key are provided with encrypter_public_key",
+			input: ClientConfigUpdate{WalletID: &walletID, EncryptedApiKey: &encryptedApiKey, EncrypterPublicKey: &encrypterPublicKey},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.input.validate()
+			if tt.wantErr != nil {
+				assert.Equal(t, tt.wantErr.Error(), err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func beginTx(t *testing.T, ctx context.Context, dbConnectionPool db.DBConnectionPool) db.DBTransaction {
+	t.Helper()
+
+	tx, err := dbConnectionPool.BeginTxx(ctx, nil)
+	require.NoError(t, err)
+
+	return tx
+}
+
+func rollback(t *testing.T, dbTx db.DBTransaction) {
+	t.Helper()
+
+	err := dbTx.Rollback()
+	require.NoError(t, err)
+}
```

Comment:
💡  These are methods we could put in the dbtest.go file.

Reply:
Maybe even a beginTx that rolls back automatically:

```suggestion
func beginTx(t *testing.T, ctx context.Context, dbConnectionPool db.DBConnectionPool, autoRollback bool) db.DBTransaction {
	t.Helper()

	tx, err := dbConnectionPool.BeginTxx(ctx, nil)
	require.NoError(t, err)
	
	if autoRollback {
		t.Cleanup(func() {
			rollback(t, dbTx)
		})
	}

	return tx
}

func rollback(t *testing.T, dbTx db.DBTransaction) {
	t.Helper()

	err := dbTx.Rollback()
	require.NoError(t, err)
}
```

Reply:
Nice, I like this suggestion 5a2b3119a9ccd6b55f17776d63e56cac176d898d